library/core/src/pin.rs

```diff
@@ -1717,10 +1718,59 @@ impl<Ptr: fmt::Pointer> fmt::Pointer for Pin<Ptr> {
 // for other reasons, though, so we just need to take care not to allow such
 // impls to land in std.
 #[stable(feature = "pin", since = "1.33.0")]
-impl<Ptr, U> CoerceUnsized<Pin<U>> for Pin<Ptr> where Ptr: CoerceUnsized<U> {}
+impl<Ptr, U> CoerceUnsized<Pin<U>> for Pin<Ptr>
+where
+    Ptr: CoerceUnsized<U> + PinCoerceUnsized,
+    U: PinCoerceUnsized,
+{
+}
+
+#[stable(feature = "pin", since = "1.33.0")]
+impl<Ptr, U> DispatchFromDyn<Pin<U>> for Pin<Ptr>
+where
+    Ptr: DispatchFromDyn<U> + PinCoerceUnsized,
+    U: PinCoerceUnsized,
+{
+}
+
+#[unstable(feature = "pin_coerce_unsized_trait", issue = "123430")]
+/// # Safety
+///
+/// Implementing this unsafe traits requires the guarantee that any two calls
+/// to `Deref::deref` must return the same value at the same address, **even after moves
+/// or unsize-coercions of `self`**, with exceptions of mutations to `self`.
+///
+/// Here, "same value" means that if `deref` returns a trait object, then the actual
+/// concrete type behind that trait object must not change.
+/// Additionally, when you unsize- coerce from `Self` to `Unsized`,
+/// then if you call `deref` on `Unsized` which returns a trait object reference,
+/// the underlying type of that trait object must be `<Self as Deref>::Target`.
+///
+/// Analogous requirements apply to other unsized types. E.g., if `deref` returns
+/// `[T]`, then the length must not change. In other words, the underlying type
+/// must not change from `[T; N]` to `[T; M]` with an `N` different from `M`.
+///
+/// If this type also implements `DerefMut`, then the same guarantee must be upheld by
+/// calls to `deref_mut`.
```

Comment:
It looks like this is using the old safety comment. The RFC was updated to use these safety requirements instead:
```
/// Trait that indicates that this is a pointer or a wrapper for one, where
/// unsizing can be performed on the pointee when it is pinned.
///
/// # Safety
///
/// If this type implements `Deref`, then the concrete type returned by `deref`
/// and `deref_mut` must not change without a modification. The following
/// operations are not considered modifications:
///
/// * Moving the pointer.
/// * Performing unsizing coercions on the pointer.
/// * Performing dynamic dispatch with the pointer.
/// * Calling `deref` or `deref_mut` on the pointer.
///
/// The concrete type of a trait object is the type that the vtable corresponds
/// to. The concrete type of a slice is an array of the same element type and
/// the length specified in the metadata. The concrete type of a sized type
/// is the type itself.
```
It gets rid of the requirement that the address is the same, and simplifies it to just requiring that the concrete type is unchanged.